tests/unit/s2n_async_pkcs11_sign_test.c

```diff
@@ -0,0 +1,574 @@
+    *out_buf = decrypted;
+    pkcs11_cleanup(session);
+
+    return S2N_SUCCESS;
+}
+
+static int pkcs11_sign(CK_SESSION_HANDLE session,
+                 CK_OBJECT_HANDLE key,
+                 uint8_t * hash_buf, 
+                 uint32_t hash_len,
+                 uint8_t ** sig_buf, 
+                 uint32_t * sig_len)
+{
+    POSIX_GUARD_PTR(hash_buf);
+    POSIX_GUARD_PTR(sig_buf);
+    POSIX_GUARD_PTR(sig_len);
+    
+
+    CK_FUNCTION_LIST_PTR function_list = NULL;
+    POSIX_GUARD(C_GetFunctionList(&function_list));
+    POSIX_GUARD_PTR(function_list);
+
+    CK_ATTRIBUTE template;
+    CK_KEY_TYPE keytype;
+    template.type = CKA_KEY_TYPE;
+    template.pValue = &keytype;
+    template.ulValueLen = sizeof(keytype);
+
+    POSIX_GUARD(function_list->C_GetAttributeValue(session, key, &template, 1));
+
+    CK_MECHANISM mechanism;
+    uint8_t * temp_digest;
+    uint32_t temp_digest_len;
+
+    if(keytype == CKK_RSA)
+    {
+        /* OpenSSL expects hashed data without padding, but PKCS #11 C_Sign function performs a hash
+         * & sign if hash algorithm is specified.  This helper function applies padding
+         * indicating data was hashed with SHA-256 while still allowing pre-hashed data to
+         * be provided. */
+        uint8_t sha256_encoding[] = pkcs11STUFF_APPENDED_TO_RSA_SIG;
+        temp_digest_len = hash_len + sizeof(sha256_encoding);
+        temp_digest = malloc(temp_digest_len);
+
+        append_sha256_id(hash_buf, temp_digest);
+        mechanism.mechanism = CKM_RSA_PKCS;
+        mechanism.pParameter= NULL;
+        mechanism.ulParameterLen = 0 ;
+    } else {
+        temp_digest = hash_buf;
+        temp_digest_len = hash_len;
+        mechanism.mechanism = CKM_ECDSA;
+        mechanism.pParameter= NULL;
+        mechanism.ulParameterLen = 0 ;
+    }
+
+    POSIX_GUARD(function_list->C_SignInit(session,
+                                           &mechanism,
+                                           key));
+
+
+    POSIX_GUARD(function_list->C_Sign(session,
+                                       temp_digest,
+                                       temp_digest_len,
+                                       NULL,
+                                       (CK_ULONG_PTR)sig_len));
+
+    /* We need an extra 8 bytes to store the ASN.1 encoding if the key is ECDSA. */
+    uint8_t * sig = malloc(*sig_len + 8);
+    POSIX_GUARD_PTR(sig);
+    POSIX_GUARD(function_list->C_Sign(session,
+                                       temp_digest,
+                                       temp_digest_len,
+                                       sig,
+                                       (CK_ULONG_PTR)sig_len));
+    if(keytype == CKK_ECDSA)
+    {
+        /* Convert from a PKCS #11 encoding to an ASN.1 encoding, sig_len will be updated to account for the extra bytes in the conversion. */
+        pkcs11_sig_to_asn1_sig(sig, sig_len);
+    }
+
+   *sig_buf = sig;
+    pkcs11_cleanup(session);
```

Comment:
Once again, why does signing terminate the connection to the token?

Also, it seems bad that connection to the token is only terminated when the operation completes successfully. If something fails, there is no way to terminate the connection.